### PR TITLE
Add Relation Extraction LoRA script and notebook tutorial

### PR DESCRIPTION
**Pull request by GauravPatil2515**
## Description
Addresses feature request #2119 by adding named entity relation classification / relation extraction examples to the repository. 

Currently, there are sequence classification and token classification examples, but no standard showcase for relation extraction between pairs of nominals. This PR implements a relation extraction tutorial using LoRA on the SemEval-2010 Task 8 dataset.

## Implementation Details
1. **Training Script** (`examples/relation_extraction/peft_lora_relation_extraction.py`): Command-line script to fine-tune `bert-base-uncased` with LoRA on relation classification.
2. **Jupyter Notebook** (`examples/relation_extraction/peft_lora_relation_extraction.ipynb`): Step-by-step interactive tutorial showcasing:
   - Adding special entity markers (`<e1>`, `</e1>`, `<e2>`, `</e2>`) to tokenizer and resizing embeddings.
   - Wrapping the sequence classifier model with `LoraConfig(task_type=TaskType.SEQ_CLS)`.
   - Custom metric evaluation using macro- and micro-F1.
   - Direct inference pipeline testing.
3. **Data Loading**: Loads the dataset from the parquet branch (`joelniklaus/sem_eval_2010_task_8` with `revision="refs/convert/parquet"`) to align with current `datasets` standards and avoid executing deprecated dataset scripts.

## Validation
I verified the implementation by running it locally; it successfully trains, runs evaluations, and saves the adapter model:
- Train/eval runs at ~19it/s.
- Validation metrics compute macro/micro F1 without errors.
- Checked styles with `make style` and `make quality`.

*Note: This PR was drafted with the assistance of an AI coding assistant (Antigravity), but all changes, execution runs, and pipeline checkpoints have been manually verified and run locally.*

### CODE REVIEW

File: examples/sequence_classification/peft_lora_relation_extraction.ipynb
Comment (BenjaminBossan) [inline examples/sequence_classification/peft_lora_relation_extraction.ipynb:1]: From my understanding, the dataset already provides the entities and the only thing left for the model is to predict what type of relationship it is dealing with. So this is in the end just a sequence classification task, right? Let's move it to `examples/sequence_classification` then.

File: examples/sequence_classification/peft_lora_relation_extraction.ipynb [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_checkpoint = \"bert-base-uncased\"\n",
+    "tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)\n",
+    "\n",
+    "special_tokens_dict = {\"additional_special_tokens\": [\"<e1>\", \"</e1>\", \"<e2>\", \"</e2>\"]}\n",
```
Comment (BenjaminBossan) [inline examples/sequence_classification/peft_lora_relation_extraction.ipynb:117]: Ouch, this is rendered as empty strings on GitHub:

Image

Please add a comment so that users are not confused.

File: examples/sequence_classification/peft_lora_relation_extraction.ipynb [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define PEFT LoRA Config for Sequence Classification\n",
+    "peft_config = LoraConfig(\n",
```
Comment (BenjaminBossan) [inline examples/sequence_classification/peft_lora_relation_extraction.ipynb:186]: As is, the newly added embedding vectors for the special tokens are not trained at all. It's not strictly necessary to train them to get good results, but it could help. Consider training them by either adding a LoRA layer for the whole embedding (which can be a bit wasteful) or by selectively training those vectors with `trainable_token_indices`.

The same applies to the script.

File: examples/sequence_classification/peft_lora_relation_extraction.py
Comment (BenjaminBossan) [inline examples/sequence_classification/peft_lora_relation_extraction.py:1]: AFAICT, this is the same training script as in the jupyter notebook. Let's add a module docstring to that effect.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for adding this notebook and script for a relation extraction task. Generally, it looks good, I just added some smaller comments.

**Comment by BenjaminBossan** [comment]
@GauravPatil2515 LMK if the PR is ready for review.

File: examples/sequence_classification/peft_lora_relation_extraction.ipynb [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+    "    model=model,\n",
+    "    args=training_args,\n",
+    "    train_dataset=tokenized_dataset[\"train\"],\n",
+    "    eval_dataset=tokenized_dataset[\"test\"],\n",
+    "    processing_class=tokenizer,\n",
+    "    data_collator=data_collator,\n",
+    "    compute_metrics=compute_metrics\n",
+    ")\n",
+    "\n",
+    "trainer.train()"
```
Comment (BenjaminBossan) [inline examples/sequence_classification/peft_lora_relation_extraction.ipynb:264]: Let's add a comment on what metrics the user can expect to see. If you have tips on how to further increase the performance, feel free to add them as well.

File: examples/sequence_classification/peft_lora_relation_extraction.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    # Start training
+    print("Starting fine-tuning with PEFT...")
+    trainer.train()
+
+    # Save model and tokenizer
+    print(f"Saving PEFT adapter to {args.output_dir}")
+    trainer.save_model(args.output_dir)
+    tokenizer.save_pretrained(args.output_dir)
+    print("Fine-tuning completed successfully!")
```
Comment (BenjaminBossan) [inline examples/sequence_classification/peft_lora_relation_extraction.py:195]: In contrast to the notebook, the script doesn't show how to load the final checkpoint for inference. This is fine because the script can be considered to be purely for training the model. But I think it would make sense to at least add a comment with how to do that. The reason is that users must remember to resize the embedding to successfully load the checkpoint, which some might forget, so it's helpful to explain that.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the updates. I just have two small comments, the rest looks good.

**Comment by BenjaminBossan** [comment]
@GauravPatil2515 Are you still on it?

**Comment by GauravPatil2515** [comment]
Addressed reviewer feedback:
- Added expected evaluation metrics (accuracy, macro/micro F1) and performance improvement tips in the tutorial notebook.
- Added example snippet in the script explaining how to load the checkpoint for inference and highlighting the need to resize token embeddings.

Ready for review @BenjaminBossan!

**Comment by GauravPatil2515** [comment]
Hi @BenjaminBossan, I have addressed both comments:
1. Added expected evaluation metrics (accuracy, macro/micro F1) and performance improvement tips in the tutorial notebook.
2. Added an inference snippet in the script explaining how to load the checkpoint and highlighting the need to resize token embeddings.

Could you please reopen the PR for review?
